test/python/utest_pymasar/test_service_config.py

```diff
@@ -0,0 +1,165 @@
+            self.assertEqual(pvgdesc, pvgroups[0][2])
+            self.assertNotEqual(None, pvgroups[0][3])
+            # The following 2 tests are to confirm the date string is in the correct format
+            self.assertEqual(3, len(pvgroups[0][3].split('-')))
+            self.assertEqual(3, len(pvgroups[0][3].split(':')))
+            self.assertEqual(None, pvgroups[0][4])
+            test_desc = 'test desc'
+            test_system = 'SR'
+            test_status = 'active'
+            test_version = 20140420
+            saveService(conn, servicename, desc=test_desc)
+
+            with self.assertRaises(Exception) as context:
+                saveServiceConfig(conn, "bad servicename", masarconf, system=test_system, status=test_status,
+                              configversion=test_version, configdesc=test_desc)
+            self.assertEqual("service with given name (bad servicename) does not exist.", context.exception.message)
+            with self.assertRaises(Exception) as context:
+                saveServiceConfig(conn, servicename, masarconf, system=test_system, status="bad status",
+                              configversion=test_version, configdesc=test_desc)
+            self.assertEqual("Service status has to be either active, or inactive", context.exception.message)
+            saveServiceConfig(conn, servicename, masarconf, system=test_system, status=test_status,
+                              configversion=test_version, configdesc=test_desc)
+            with self.assertRaises(Exception) as context:
+                saveServiceConfig(conn, servicename, masarconf, system=test_system, status=test_status,
+                                  configversion=test_version, configdesc=test_desc)
+            self.assertEqual('service config exists already.', context.exception.message)
+
+            with self.assertRaises(IndexError):  # TODO: Should this be a more specific error message?
+                saveServicePvGroup(conn, "bad config", [pvgname])
+            with self.assertRaises(Exception) as context:
+                res3 = saveServicePvGroup(conn, masarconf, ["bad pvgname"])  # this test prints a message to console
+            self.assertEqual("given pv group name (bad pvgname) does not exist.", context.exception.message)
+            res3 = saveServicePvGroup(conn, masarconf, [pvgname])
+            self.assertEqual([1], res3)
+            with self.assertRaises(Exception) as context:
+                pvlist = retrieveServiceConfigPVs(conn, masarconf, servicename="bad servicename")
+            pvlist = retrieveServiceConfigPVs(conn, masarconf, servicename=servicename)
+            self.assertEqual(context.exception.message, 'Given service (bad servicename) does not exist.')
+            self.assertEqual(pvs, pvlist)
+            results = retrieveServiceConfigs(conn, "bad servicename", masarconf)
+            self.assertEqual(1, len(results))  # IE: no data returned
```

Comment:
I find it helpful to test list lengths last.  If the length is too short then another test will fail first, probably with more informative output.  So the length test just catches the too-long case.